### PR TITLE
IBFT Documentation

### PR DESCRIPTION
**Pull request by jbhurat**
Added documentation for IBFT including Overview and Implementation

### CODE REVIEW

**Review by fixanoid — DISMISSED** [review]

File: docs/Consensus/ibft/ibft.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    - Otherwise, it picks `1 + current round number` as the proposed round number.
+- Whenever a validator receives `F + 1` of `ROUND CHANGE` messages on the same proposed round number, it compares the received one with its own. If the received is larger, the validator broadcasts `ROUND CHANGE` message again with the received number.
+- Upon receiving `ceil(2N/3)` of `ROUND CHANGE` messages on the same proposed round number, the **validator** exits the round change loop, calculates the new **proposer**, and then enters `NEW ROUND` state.
+- Another condition that a validator jumps out of round change loop is when it receives verified block(s) through peer synchronization.
+
+#### Proposer selection
+
+Currently we support two policies: **round robin** and **sticky proposer**.
+
+- Round robin: in a round robin setting, proposer will change in every block and round change.
```
Comment (drequinox) [inline docs/Consensus/ibft/ibft.md]: IMO it's worth mentioning that RoundRobin is the default policy.

File: docs/Consensus/Consensus.md [hunk 2]
```diff
@@ -5,7 +5,7 @@ With no need for POW/POS in a permissioned network, Quorum instead offers multip
 * __Raft-based Consensus__: A consensus model for faster blocktimes, transaction finality, and on-demand block creation.  See [Raft-based consensus for Ethereum/Quorum](../raft) for more information 
 
 
-* __Istanbul BFT (Byzantine Fault Tolerance) Consensus__: A PBFT-inspired consensus algorithm with transaction finality, by AMIS.  See [Istanbul BFT Consensus documentation](https://github.com/ethereum/EIPs/issues/650), the [RPC API](../ibft/istanbul-rpc-api.md), and this [technical web article](https://medium.com/getamis/istanbul-bft-ibft-c2758b7fe6ff) for more information
+* __Istanbul BFT (Byzantine Fault Tolerance) Consensus__: A PBFT-inspired consensus algorithm with transaction finality, by AMIS.  See [Istanbul BFT Consensus documentation](../ibft/ibft), the [RPC API](../ibft/istanbul-rpc-api), and this [technical web article](https://medium.com/getamis/istanbul-bft-ibft-c2758b7fe6ff) for more information
```
Comment (drequinox) [inline docs/Consensus/Consensus.md]: should we say "immediate transaction finality", as other algorithms do have some notion of finality too?

File: docs/Consensus/ibft/ibft.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+- `Validator`: Block validation participant.
+- `Proposer`: A block validation participant that is chosen to propose block in a consensus round.
+- `Round`: Consensus round. A round starts with the proposer creating a block proposal and ends with a block commitment or round change.
+- `Proposal`: New block generation proposal which is undergoing consensus processing.
+- `Sequence`: Sequence number of a proposal. A sequence number should be greater than all previous sequence numbers. Currently each proposed block height is its associated sequence number.
+- `Backlog`: The storage to keep future consensus messages.
+- `Round state`: Consensus messages of a specific sequence and round, including pre-prepare message, prepare message, and commit message.
+- `Consensus proof`: The commitment signatures of a block that can prove the block has gone through the consensus process.
+- `Snapshot`: The validator voting state from last epoch.
+
```
Comment (drequinox) [inline docs/Consensus/ibft/ibft.md:20]: please add, epoch definition: 
The number of blocks (default 30000 blocks) after which to checkpoint and reset the pending votes.

File: docs/Consensus/ibft/ibft.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+#### Proposer selection
+
+Currently we support two policies: **round robin** and **sticky proposer**.
+
+- Round robin: in a round robin setting, proposer will change in every block and round change.
+- Sticky proposer: in a sticky proposer setting, proposer will change only when a round change happens.
+
+#### Validator list voting
+
+Istanbul BFT uses a similar validator voting mechanism as Clique and copies most of the content from Clique [EIP](https://github.com/ethereum/EIPs/issues/225). Every epoch transaction resets the validator voting, meaning if an authorization or de-authorization vote is still in progress, that voting process will be terminated.
```
Comment (drequinox) [inline docs/Consensus/ibft/ibft.md]: I think "every epoch transaction .... " is not very clear. something like 
"every epoch any pending votes are reset". 
is more accurate.

File: docs/Consensus/ibft/ibft.md (hunk 1 above)
Comment (jbhurat) [reply docs/Consensus/ibft/ibft.md]: Updated the documentation to reflect this

File: docs/Consensus/Consensus.md (hunk 2 above)
Comment (jbhurat) [reply docs/Consensus/Consensus.md]: Updated the documentation to reflect this

File: docs/Consensus/ibft/ibft.md (hunk 4 above)
Comment (jbhurat) [reply docs/Consensus/ibft/ibft.md]: Updated the documentation to reflect this

File: docs/Consensus/ibft/ibft.md (hunk 3 above)
Comment (jbhurat) [reply docs/Consensus/ibft/ibft.md:20]: EPOCH definition was already present updated documentation to add default value

**Review by fixanoid — APPROVED** [review]

**Pull request merged**